bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -252,6 +252,48 @@ impl Connection {
         Ok(self)
     }
 
+    /// Signals the connection to do a key_update at the next possible opportunity.
+    /// Note that the resulting key update message will not be sent until `send` is
+    /// called on the connection.
+    ///
+    /// `peer_request` indicates if a key update should also be requested
+    /// of the peer. When set to `KeyUpdateNotRequested`, then only the sending
+    /// key of the connection will be updated. If set to `KeyUpdateRequested`, then
+    /// the sending key of conn will be updated AND the peer will be requested to
+    /// update their sending key. Note that s2n-tls currently only supports
+    /// `peer_request` being set to `KeyUpdateNotRequested` and will return an error
+    /// if any other value is used.
+    pub fn request_key_update(&mut self, peer_request: PeerKeyUpdate) -> Result<&mut Self, Error> {
+        unsafe {
+            s2n_connection_request_key_update(self.connection.as_ptr(), peer_request.into())
+                .into_result()
+        }?;
+        Ok(self)
+    }
+
+    /// Reports the number of times sending and receiving keys have been updated.
+    ///
+    /// This only applies to TLS1.3. Earlier versions do not support key updates.
+    ///
+    /// s2n-tls only tracks up to u8::MAX (255) key updates. If this method
+    /// reports 255 updates, then more than 255 updates may have occurred.
+    ///
+    /// The return value is a tuple of `(send_key_updates, recv_key_updates)`
+    #[cfg(feature = "unstable-ktls")]
+    pub fn key_update_counts(&self) -> Result<(u8, u8), Error> {
```

Comment:
You just want this for testing, right? I wonder if there's any way we can avoid making it completely public. I wrote it up pretty quickly and even now I'm tempted to bump its arguments up to something bigger than a u8. I'm skeptical the C method won't change before stabilizing.

Reply:
Yup, I'm just using it for testing.

> if there's any way we can avoid making it completely public.

Hmm. Currently it's gated behind the `unstable-ktls` feature, so a customer has to explicitly opt in to using it. This matches the "stability" that we've used with other unstable methods, like https://github.com/aws/s2n-tls/blob/30a490170d9e560abaf9a38de8851b846aef479b/bindings/rust/s2n-tls/src/connection.rs#L755-L762

My personal take is that customers opting to use the "unstable-ktls" feature are obviously opting into some breakage, which is the benefit of "unstable" headers.

If we're uncomfortable with the proposed level of exposure in the bindings, I'd probably just nuke them entirely and directly call the underlying `s2n_connection_get_key_update_counts` from rust.

Reply:
I'm probably just being too cautious. I'm fine with it as-is.